### DISCUSSION TITLE
ALT+D duplicate current ssh like putty

### DISCUSSION THREAD

**Discussion by yar999** [wezterm/wezterm#1937] [Ideas]
# `ALT+D` duplicate current ssh like putty

**The implementation**:

1. `ssh.bat` calls Python to set the parameters of the SSH call
2. Wezterm gets the ssh call parameters we set earlier via pane:get_user_vars()
3. `Alt + D` calls the custom event of EmitEvent

**Imperfect place**

After SSH exits, `Alt + D` in the current TAB can still call the SSH parameters set before.

Let me know if you have a better idea.

`d:\tools\ssh.bat`:

```
@echo off

python d:\tools\printf.py "%*"

wsl ssh %*
```

`d:\tools\printf.py`:

```python
from sys import argv
from base64 import b64encode

be=b64encode(" ".join(argv[1:]).encode()).decode()
print(f"\033]1337;SetUserVar=cssh={be}\007",end="")
```

`wezterm.lua`:

```lua
...

local function pcssh(cssh)
    local args = {}
    table.insert(args, "cmd.exe")
    table.insert(args, "/k")
    table.insert(args, "ssh")
    for i in cssh:gmatch("([^%s]+)") do
        table.insert(args,i)
    end
    return args
end

wezterm.on("duplicate-current-ssh",function(window, pane)
    local cssh = pane:get_user_vars().cssh
    if cssh then
        wezterm.log_info(cssh)
        window:perform_action(wezterm.action{SpawnCommandInNewTab={
            args=pcssh(cssh)}
        }, pane)
    end
end)

return {
    ...
    {key="d", mods="ALT", action=wezterm.action{EmitEvent="duplicate-current-ssh"}},
    ...
}
```

**Comment by wez** [comment]
FWIW, your post made me think of https://github.com/wez/wezterm/issues/580 which, at the very bottom, is asking for wezterm to support re-spawning a finished process

**Reply by yar999** [reply]
My post idea can solve #580 to some extent?

Thanks a lot to the wezterm project, we can flexibly configure to meet our own requirements. @wez

**Comment by yar999** [comment]
SetUserVar + SendString + SendKey  can implement  #580 restart the closed process (with the same params).

Reconnect the disconnected server using ssh:

```lua
wezterm.on("reconnect-ssh",function(window, pane)
    local cssh = pane:get_user_vars().cssh
    if cssh ~= "" and cssh ~= nil then
        -- wezterm.log_info(cssh)
        window:perform_action(wezterm.action{SendString="ssh " .. cssh
        }, pane)
        window:perform_action(wezterm.action{SendKey={
            key="Enter"}
        }, pane)
    end
end)

return {
    ...
    {key="r", mods="ALT", action=wezterm.action{EmitEvent="reconnect-ssh"}},
    ...
}
```
